subsys/bluetooth/audio/aics.c

```diff
@@ -0,0 +1,824 @@
+	}
+
+	CHECKIF(!init) {
+		BT_DBG("NULL init");
+		return -EINVAL;
+	}
+
+	if (!instance_prepared) {
+		prepare_aics_instances();
+		instance_prepared = true;
+	}
+
+	CHECKIF(aics->srv.initialized) {
+		return -EALREADY;
+	}
+
+	CHECKIF(init->mute > BT_AICS_STATE_MUTE_DISABLED) {
+		BT_DBG("Invalid AICS mute value: %u", init->mute);
+		return -EINVAL;
+	}
+
+	CHECKIF(init->gain_mode > BT_AICS_MODE_AUTO) {
+		BT_DBG("Invalid AICS mode value: %u", init->gain_mode);
+		return -EINVAL;
+	}
+
+	CHECKIF(init->type > BT_AICS_INPUT_TYPE_STREAMING) {
+		BT_DBG("Invalid AICS input type value: %u", init->type);
+		return -EINVAL;
+	}
+
+	CHECKIF(init->units == 0) {
+		BT_DBG("AICS units value shall not be 0");
+		return -EINVAL;
+	}
+
+	CHECKIF(!(init->min_gain <= init->max_gain)) {
+		BT_DBG("AICS min gain (%d) shall be lower than or equal to max gain (%d)",
+		       init->min_gain, init->max_gain);
+		return -EINVAL;
+	}
+
+	CHECKIF(init->gain < init->min_gain || init->gain > init->max_gain) {
+		BT_DBG("AICS gain (%d) shall be not lower than min gain (%d) "
+		       "or higher than max gain (%d)",
+		       init->gain, init->min_gain, init->max_gain);
+		return -EINVAL;
+	}
+
+	aics->srv.state.gain = init->gain;
+	aics->srv.state.mute = init->mute;
+	aics->srv.state.gain_mode = init->gain_mode;
+	aics->srv.gain_settings.units = init->units;
+	aics->srv.gain_settings.minimum = init->min_gain;
+	aics->srv.gain_settings.maximum = init->max_gain;
+	aics->srv.type = init->type;
+	aics->srv.status = init->status ? BT_AICS_STATUS_ACTIVE : BT_AICS_STATUS_INACTIVE;
+
+	if (init->description) {
+		strncpy(aics->srv.description, init->description,
+			sizeof(aics->srv.description) - 1);
+		/* strncpy may not always null-terminate */
+		aics->srv.description[sizeof(aics->srv.description) - 1] = '\0';
+		if (IS_ENABLED(CONFIG_BT_DEBUG_AICS) &&
+		    strcmp(aics->srv.description, init->description)) {
+			BT_DBG("Input desc clipped to %s",
+			       log_strdup(aics->srv.description));
+		}
+	}
+
+	/* Iterate over the attributes in AICS (starting from i = 1 to skip the
+	 * service declaration) to find the BT_UUID_AICS_DESCRIPTION and update
+	 * the characteristic value (at [i]), update that with the write
+	 * permission and callback, and also update the characteristic
+	 * declaration (always found at [i - 1]) with the
+	 * BT_GATT_CHRC_WRITE_WITHOUT_RESP property.
+	 */
+	if (init->desc_writable) {
+		for (int i = 1; i < aics->srv.service_p->attr_count; i++) {
+			struct bt_gatt_attr *attr;
+
+			attr = &aics->srv.service_p->attrs[i];
+
+			if (!bt_uuid_cmp(attr->uuid, BT_UUID_AICS_DESCRIPTION)) {
+				/* Update attr and chrc to be writable */
+				struct bt_gatt_chrc *chrc;
+
+				chrc = aics->srv.service_p->attrs[i - 1].user_data;
+				attr->write = write_description;
+				attr->perm |= BT_GATT_PERM_WRITE_ENCRYPT;
+				chrc->properties |= BT_GATT_CHRC_WRITE_WITHOUT_RESP;
+
+				break;
+			}
+		}
+	}
+
+	err = bt_gatt_service_register(aics->srv.service_p);
```

Comment:
Is this calling register the instances from bt_aics_free_instance_get? Im not sure how that would work because those instances are already registered in the static portion of the table so this would register it again.

Reply:
The instances are defined with `BT_GATT_SERVICE_INSTANCE_DEFINE` which doesn't statically register the GATT services, so this should work correctly.